base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs

```diff
@@ -154,275 +154,427 @@ where TBackend: TransactionBackend + 'static
+                        self.tx_id
+                    );
+                    return Err(TransactionServiceProtocolError::new(self.tx_id, TransactionServiceError::Shutdown))
+                },
+            }
 
-                if fetch_utxo_response_received {
-                    if let Some(result) = chain_metadata_response_received {
-                        // If the tip has moved beyond this Coinbase transaction's blockheight and it wasn't mined then
-                        // it should be cancelled
-
-                        if !result {
-                            error!(
+            let delay = delay_for(self.resources.config.peer_dial_retry_timeout);
+            let mut base_node_connection = match connection {
+                None => {
+                    futures::select! {
+                        _ = delay.fuse() => {
+                            continue;
+                        },
+                        _ = shutdown => {
+                            info!(
                                 target: LOG_TARGET,
-                                "Chain tip has moved ahead of this Coinbase transaction's block height without it \
-                                 being mine. Cancelling Coinbase transaction (TxId: {})",
-                                completed_tx.tx_id
+                                "Coinbase Monitoring Protocol (TxId: {}) shutting down because it received the \
+                                 shutdown signal",
+                                self.tx_id
                             );
-                            if let Err(e) = self
-                                .resources
-                                .output_manager_service
-                                .cancel_transaction(completed_tx.tx_id)
-                                .await
-                            {
-                                warn!(
+                            return Err(TransactionServiceProtocolError::new(
+                                self.tx_id,
+                                TransactionServiceError::Shutdown
+                            ))
+                        },
+                    }
+                },
+                Some(c) => c,
+            };
+            let mut client = match base_node_connection
+                .connect_rpc_using_builder(
+                    BaseNodeWalletRpcClient::builder().with_deadline(self.resources.config.chain_monitoring_timeout),
+                )
+                .await
+            {
+                Ok(c) => c,
+                Err(e) => {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Problem establishing RPC connection (TxId: {}): {}", self.tx_id, e
+                    );
+                    delay.await;
+                    continue;
+                },
+            };
+
+            let signature = completed_tx.transaction.body.kernels()[0].clone().excess_sig;
```

Comment:
I know that the body is supposed to have at least 1 kernel, but this will panic instead of failing gracefully if that's not true. Do we already have mitigations in place that ensure the body will **always** have at least 1 kernel?

Reply:
Good point - Implemented safe code.